### ISSUE TITLE
Windows Installer not saving exult.cfg properly

### ISSUE THREAD

**Issue by wench** [exult/exult#432]
### From Discussion https://github.com/exult/exult/discussions/431

<div type='discussions-op-text'>

<sup>Originally posted by **2bitcounter** January 15, 2024</sup>
Hello,

I bought GOG.com's "Ultima 7 The Complete Collection" and installed each game's installer separately for U7(BG+Forge) and U7:SI+SS. I am using Win10 64-bit.
I then installed Exult v1.8.0 and did not change any setup options in the installer (i.e. it will setup games).
It asked for the paths for both games, which were 
"C:\GOG Games\Ultima 7" and "C:\GOG Games\Ultima 7 - Serpent Isle". Note that the Exult installer did not complain about either of the paths, and I was able to complete the Exult install.
However, on running Exult, it complained about "could not find the static folder" for both games.

Here is what exult.cfg looked after installing and before any editing was done:
<config>
 <disk>
  <save_compression_level>
  1
  </save_compression_level>
 </disk>
 <video>
  <fullscreen>
  no
  </fullscreen>
  <game>
   <border>
    <red>
    0
    </red>
    <green>
    0
    </green>
    <blue>
    0
    </blue>
   </border>
   <width>
   320
   </width>
   <height>
   200
   </height>
  </game>
  <display>
   <width>
   1024
   </width>
   <height>
   768
   </height>
  </display>
  <scale>
  2
  </scale>
  <scale_method>
  Point
  </scale_method>
  <fill_mode>
  Fit
  </fill_mode>
  <fill_scaler>
  Point
  </fill_scaler>
  <highdpi>
  yes
  </highdpi>
  <share_video_settings>
  yes
  </share_video_settings>
 </video>
 <gameplay>
  <right_click_closes_gumps>
  yes
  </right_click_closes_gumps>
  <gumps_dont_pause_game>
  no
  </gumps_dont_pause_game>
  <mouse3rd>
  no
  </mouse3rd>
  <fastmouse>
  no
  </fastmouse>
  <double_click_closes_gumps>
  no
  </double_click_closes_gumps>
  <combat>
   <difficulty>
   0
   </difficulty>
   <charmDifficulty>
   normal
   </charmDifficulty>
   <mode>
   original
   </mode>
   <show_hits>
   no
   </show_hits>
  </combat>
  <step_tile_delta>
  8
  </step_tile_delta>
  <allow_right_pathfind>
  double
  </allow_right_pathfind>
  <formation>
  yes
  </formation>
  <smooth_scrolling>
  0
  </smooth_scrolling>
  <alternate_drop>
  no
  </alternate_drop>
  <allow_autonotes>
  no
  </allow_autonotes>
  <scroll_with_mouse>
  yes
  </scroll_with_mouse>
  <extended_intro>
  yes
  </extended_intro>
 </gameplay>
 <audio>
  <disablepause>
  no
  </disablepause>
  <enabled>
  yes
  </enabled>
  <midi>
   <enabled>
   yes
   </enabled>
   <convert>
   gm
   </convert>
   <use_oggs>
   no
   </use_oggs>
   <driver>
   gm
   </driver>
  </midi>
 </audio>
 <touch>
  <item_menu>
  no
  </item_menu>
  <dpad_location>
  no
  </dpad_location>
  <touch_pathfind>
  no
  </touch_pathfind>
 </touch>
 <shortcutbar>
  <use_shortcutbar>
  no
  </use_shortcutbar>
  <use_outline_color>
  no
  </use_outline_color>
  <hide_missing_items>
  yes
  </hide_missing_items>
 </shortcutbar>
</config>

In case the above copy & paste looks like garbage I'm also attaching the original non-working .cfg file to this message.

`You'll notice that despite the installer being pointed to the GOG file paths there is no <blackgate> or  <serpentisle> with the relevant <path> sections under <games>`

I am not sure why this happened, as the GOG path in Exult docs section 11.9 matches my path for both games:
_Right now the only way to get at the game files is to run the installer on Windows (or in WINE if you are not a Windows user). After the installation you will find the game files in C:\GOG Games\Ultima 7 and C:\GOG Games\Ultima 7 - Serpent Isle.
After installation you can just run our installer and when asked where the games are installed enter or browse to the above paths.
Unfortunately GOG changes the download of the games for some unknown reason again and again. So our instructions are sometimes out of date. These ones are for the latest GOG release at the time of writing. If it doesn't work, please visit our forum and tell us._ 

So I'm telling you ;-)

**The way I got both games to work was by replacing the contents of my exult.cfg file in %APPDATA%/Local/Exult with the one from the following link:
[https://pastebin.com/zR6wNgHk](https://pastebin.com/zR6wNgHk)
and then editing exult.cfg and:**
```
changing the path for <blackgate> to "C:\GOG Games\Ultima 7"
changing the path for <serpentisle> to "C:\GOG Games\Ultima 7 - Serpent Isle"
```

I do not know why the Exult installer is not setting up the paths or other options in exult.cfg correctly (there's a lot of stuff in the pastebin link that was not in my .cfg after installing).

But it was worth this hassle - Ultima looks great with the antialias filtering and much better than the original DOS version (running under DOSBox).

I have documented this so it can be fixed, and also to save someone who reads this in the future trying to figure out why it ain't working.

[exult - Original Copy.zip](https://github.com/exult/exult/files/13934483/exult.-.Original.Copy.zip)
</div>

**Comment by 2bitcounter** [comment]
I tried installing Exult 1.8.0 (using same install file that had problems as described before) to a Win7 Pro 64-bit test computer - it worked after install and was able to run Ultima 7 BG without issue or need to manually modify anything.
The Win7 PC was logged in as an admin user, the Win10 was a non-admin user that was prompted for Admin credentials during Exult install (not sure if that makes a difference, but I see that exconfig.cc helps the installer find paths etc).

By the way, someone else had the same issue: 
[https://old.reddit.com/r/Ultima/comments/16wze7r/exult_cant_find_ultima_7_files/](https://old.reddit.com/r/Ultima/comments/16wze7r/exult_cant_find_ultima_7_files/)
This is where I found the pastebin contents of exult.cfg

**Comment by wench** [comment]
Administrator privileges should not be an issue as exult.cfg if written to the user's App data directory, not anywhere that need Administrator privileges. Windows version should not matter either unless there was some change in innosetup that alters bahaviouys on different versions of Windows. Exconfig was originally written for a much older version of innosetup than is being used now. My suspicion is innosetup changed something and it causes an intermittent fault with exconfig.

**Comment by KnightCaptainU7** [comment]
The handling of exult.cfg is a pain, as it gets overwritten even if no changes are made to it.

**Comment by DominusExult** [comment]
> The handling of exult.cfg is a pain, as it gets overwritten even if no changes are made to it.

Hmm, AFAICT the only parts of exult.cfg that gets overwritten are the path settings. Other (or most?) settings are left untouched when you reinstall exult (tested right now with the 1.8 release).

**Comment by wench** [comment]
I added some extra code to attempt to ensure that it doesn't try to change the paths if they are not valid but the installer will still rewrite the config file even if there are no changes but this should not be a problem as the exult.cfg code is shared by the installer and exult itself so if there is a problem saving exult.cfg then exult itself on Windows should exhibit the same problem so I am fairly confident there is no problem with resaving an unchanged exult.cfg

**Comment by 2bitcounter** [comment]
Note that the original issue discussed was for a clean install (exult never previously installed) on a Win10 PC. As for rewriting the .cfg file, from my earlier post:
`"despite the installer being pointed to the GOG file paths there is no <blackgate> or  <serpentisle> with the relevant <path> sections under <games>"`

Is there anything that would cause the installer to NOT have the above sections under games?

**Comment by wench** [comment]
The installer is supposed to verify the paths before letting the user progress. However the settings are saved much later than the paths are verified so if the values of the paths changed between being checked and written then exult.cfg would be incorrect. As exult.cfg is xml like if the value being set is whitespace that is the same as empty. And empty keys will not written out. So, if for some reason the paths got set to white space this behaviors might be explainable in the older version of the installer. As it is right now the installer will not try to set a path to whitespace.

As for why the path might change, the installer used the values of the UI EditBoxes elements for the paths. There might be some reason why the EditBoxes could have their value changed after the installer verifies the paths before it writes them. I just changed things so the installer grabs the EditBox values and stores it elsewhere once they are verified so if the textboxes can change it will no longer be a problem. Hopefully all my changes fix this problem but it's hard to know without reproducing it to test. I just fixed things that are not as robust as I think they should be

**Comment by wench** [comment]
My suspicion is that this might be a race condition where at the end of installation when exult.cfg is saved, innosetup starts cleaning up memory and clears the values of text boxes. If this is the case then why did this ever work.

**Comment by wench** [comment]
I have actually managed to sort of reproduce this. Installing a snapshot and choosing not to setup game paths seems to be trashing my Serpent Isle Paths sometimes. Some more testing shown me that the snapshot from march 4th does not exhibit the problem I can only reproduce it reliably for snapshots from the 19th and after. Seems that my changes from the 18th to exconfig has made this worse, not better

**Comment by wench** [comment]
Ok i fixed the problem i just commented about but It is not the original problem of this issue but the fixes in this latest commit should hopefully prevent the problem which is what my previous broken commit was supposed to do. Leaving this open in case something else comes up.

**Comment by wench** [comment]
Actually I am going to close this as my latest commit fixed a bug in the iss file that had been there since the beginning and likely was the cause of the original problem the user experienced. Two strings in CurPageChanged were incorrectly set to the String type instead of AnsiString when they were passed to exconfig.dll

I tested the v1.7.0.20220415 snapshot and it exhibits the exact problem I just fixed and trashes both the BG and SI paths if the option to set game paths is not selected. I suspect the exact behaviour of the bug is impacted by the version of innosetup and commit messages hint that the problem likely only became an issue when we started using innosetup 6.0.4 as the same problem was fixed in other functions for that version.
